### PR TITLE
RUN-4373 Window groups to Typescript

### PR DESCRIPTION
**Pull request by pbaize**
ℹ️ About
This PR transitions window_groups handler from JavaScript to TypeScript

⚠️ Yes, used git mv to rename the file.

🚥 Automated test results
✅ [Windows 7](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/5b72003454b21953031f396b)
✅ [Windows 10](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/5b72015b54b21953031f396c)
(Backwards-compatible App.getInfo probably needs a revised timeout for hosting the config and downloading the old runtime)

### CODE REVIEW

File: src/browser/window_groups.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+function mapEventWindowGroups(group: OpenFinWindow[]) {
+    return _.map(group, (win) => {
+        return {
+                appUuid: win.app_uuid,
+                windowName: win.name
+        };
+    });
+}
+const windowGroups = new WindowGroups();
+export default windowGroups;
```
Comment (whyn07m3) [inline src/browser/window_groups.ts]: Can it be a one-liner `export default new WindowGroups();`?

File: src/browser/window_groups.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+import { OpenFinWindow } from '../shapes';
+
+let uuidSeed = 0;
+
+class WindowGroups extends EventEmitter {
+    constructor() {
+        super();
+    }
+    private _windowGroups: { [groupName: string]: { [windowName: string]: OpenFinWindow; } } = {};
+    public getGroup = (uuid: string) => {
```
Comment (whyn07m3) [inline src/browser/window_groups.ts]: Missing return shape

File: src/browser/window_groups.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            delete this._windowGroups[groupUuid];
+        }
+    };
+}
+
+
+// Helpers
+
+function generateUuid(): string {
+    return 'group' + (uuidSeed++);
```
Comment (whyn07m3) [inline src/browser/window_groups.ts]: Optional, can use template literal

File: src/browser/window_groups.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        delete this._windowGroups[sourceGroupUuid];
+    };
+
+    public _addWindowToGroup = (uuid: string, win: OpenFinWindow): string => {
+        const _uuid = uuid || generateUuid();
+        this._windowGroups[_uuid] = this._windowGroups[_uuid] || {};
+        this._windowGroups[_uuid][win.name] = win;
+        return _uuid;
+    };
+
```
Comment (licui3936) [inline src/browser/window_groups.ts]: It should be private function.

File: src/browser/window_groups.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            this._removeWindowFromGroup(groupUuid, lastWindow);
+            this.emit('group-changed', {
+                groupUuid,
+                payload: generatePayload('disband', lastWindow, lastWindow, [], [])
+            });
+            lastWindow.groupUuid = null;
+            delete this._windowGroups[groupUuid];
+        }
+    };
+}
```
Comment (licui3936) [inline src/browser/window_groups.ts]: same for above two functions.

**Pull request merged**